Avoid doing event-related work for CSS animation events when no event listener is registered

cdumez opened the pull request:
#### f3a4982bb1a712a260572205d199b16147ce6797
<pre>
Avoid doing event-related work for CSS animation events when no event listener is registered
<a href="https://bugs.webkit.org/show_bug.cgi?id=254546">https://bugs.webkit.org/show_bug.cgi?id=254546</a>

Reviewed by Antoine Quint.

* Source/WebCore/animation/DeclarativeAnimation.cpp:
(WebCore::DeclarativeAnimation::invalidateDOMEvents):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::addListenerTypeIfNeeded):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/EventNames.h:
(WebCore::EventNames::isCSSAnimationEventType const):

Canonical link: <a href="https://commits.webkit.org/262231@main">https://commits.webkit.org/262231@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/25d060cbdb94f81e3a018a03200bbbeac2c64fa6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/937 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/963 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1001 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1364 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/831 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/926 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1015 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1049 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1042 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/946 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/886 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/881 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1287 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/931 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/878 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/884 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/854 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/905 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/1922 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/895 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/849 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/842 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/887 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/233 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/909 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

graouts reviewed (approved):
I'm not a fan of the dissymmetry in `DeclarativeAnimation::invalidateDOMEvents()` which is now structured in that way:

```c++
if (is<CSSAnimation>(*this)) {
    if (m_owningElement->document().hasListenerType(Document::CSS_ANIMATION_LISTENER)) {
        // enqueue animation events
    }
} else if (is<CSSTransition>(*this) && m_owningElement->document().hasListenerType(Document::TRANSITION_ANIMATION_LISTENER)) {
    // enqueue transition events
}
```

I'm thinking we should introduce a lambda for animation events and one for transition events to reduce the nesting here. Something like:

```c++
auto invalidateCSSAnimationEventsIfNeeded = [&]() {
    if (!is<CSSAnimation>(*this))
        return;
    if (!m_owningElement->document().hasListenerType(Document::CSS_ANIMATION_LISTENER))
        return;
    // enqueue animation events
};

auto invalidateCSSTransitionEventsIfNeeded = [&]() {
    if (!is<CSSTransition>(*this))
        return;
    if (!m_owningElement->document().hasListenerType(Document::CSS_TRANSITION_LISTENER))
        return;
    // enqueue transition events
};

invalidateCSSAnimationEventsIfNeeded();
invalidateCSSTransitionEventsIfNeeded();
```

Patch is obviously fine to land as-is, but this could be a nice improvement that can be done as a follow-up as well.

cdumez committed:
Avoid doing event-related work for CSS animation events when no event listener is registered
https://bugs.webkit.org/show_bug.cgi?id=254546

Reviewed by Antoine Quint.

* Source/WebCore/animation/DeclarativeAnimation.cpp:
(WebCore::DeclarativeAnimation::invalidateDOMEvents):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::addListenerTypeIfNeeded):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/EventNames.h:
(WebCore::EventNames::isCSSAnimationEventType const):

Canonical link: https://commits.webkit.org/262231@main
